File: C++/Primeros_Pasos_en_C++/Conjuntos_Disjuntos/conjuntos_disjuntos.cpp
#include <bits/stdc++.h>

using namespace std;

const int N = 1e5 +10;

int p[N], sz[N];   //  Disjoint Set Union (Union Find)

void init(int n){
	for(int i = 1; i <= n; i++){
		p[i]=i;
		sz[i]=1;
	}
}

int _find2(int u){
	if(u == p[u]) return u;
	return _find2(p[u]);
}

int _find(int u){
	if(u == p[u]) return u;
	return p[u] = _find(p[u]); // Memorizacion
}

void _union(int u,int v){
	int pu = _find(u);
	int pv = _find(v);
	if(pu == pv) return;
	p[pu] = pv;
	sz[pv] = sz[pv] + sz[pu];
}
int main(){
	// Numero de elementos
	cout << "Nro de elementos:  ";
	int  n;
	cin >> n;
	init(n);
	cout << "Nro de preguntas:  ";
	// Union de conjuntos
	// Verificar si dos elementos pertenecen al mismo conjunto
	int q;
	cin>>q;
	while(q--){
		int type;
		cin >> type;
		if(type == 1){
			int u ,v;
			cin >>u >>v ;
			_union(u,v);
		}else if(type == 2){
		// Verificar si estan en el mismo conjunto
			int u ,v;
			cin >> u >> v;
			if (_find(u) == _find(v)){
				cout << "Yes\n";
			} else {
				cout << "No\n";
			}
		}else{
			// Tamaño del conjunto
			int u ;
			cin >> u;
			int pu = _find(u);
			cout << "Tamaño:  " << sz[pu] <<"\n";
		}
		
	}
	return 0;
}

File: C++/Primeros_Pasos_en_C++/Conjuntos_Disjuntos/conjuntos_disjuntos.py
########################################
# Implementacion de conjuntos disjuntos
########################################

p = []
sz = []

def init(n):
	p = []
	sz = []
	for i in range(n + 1):
		p.append(i)
		sz.append(1)
	return p ,sz

def _find2(u):
	if u == p[u]:
		return u
	return _find2(p[u])

def _find(u):
	if u == p[u]:
		return u
	p[u] = _find(p[u])
	return p[u]

def _union(u, v):
	pu = _find(u)
	pv = _find(v)
	if pu == pv:
		return
	p[pu] = pv
	sz[pv] = sz[pv] + sz[pu]

print('Numero de elementos:  ',end = "")
n = int(input())
# Iniciamos las variables
p, sz = init(n)

print('Numero de preguntas:  ',end = '')
q=int(input())
while q > 0:
	line = input().split()
	_type = int(line[0])
	if _type == 1:
		# Union de conjuntos
		u = int(line[1])
		v = int(line[2])
		_union(u, v)
	elif _type == 2:
		# Verificar si estan en el mismo conjunto
		u = int(line[1])
		v = int(line[2])
		if _find(u) == _find(v):
			print('Yes')
		else:
			print('No')
	else:
		# Tamaño del conjunto
		u = int(line[1])
		print('Tamaño:  ',sz[_find(u)])
q =  q-1